src/span.h

```diff
@@ -21,6 +21,62 @@
 /** A Span is an object that can refer to a contiguous sequence of objects.
  *
  * It implements a subset of C++20's std::span.
+ *
+ * Things to be aware of when writing code that deals with Spans:
+ *
+ * - Similar to references themselves, Spans are subject to reference lifetime
+ *   issues. The user is responsible for making sure the objects pointed to by
+ *   a Span live as long as the Span is used. For example:
+ *
+ *       std::vector<int> vec{1,2,3,4};
+ *       Span<int> sp(vec);
+ *       vec.push_back(5);
+ *       printf("%i\n", sp.front()); // UB!
+ *
+ *   may exhibit undefined behavior, as increasing the size of a vector may
+ *   invalidate references.
+ *
+ * - One particular pitfall is that Spans can be constructed from temporaries,
+ *   but this is unsafe when the Span is stored in a variable, outliving the
+ *   temporary. For example, this will compile, but exhibits undefined behavior:
+ *
+ *       Span<const int> sp(std::vector<int>{1, 2, 3});
+ *       printf("%i\n", sp.front()); // UB!
```

Comment:
Turns out clang actually has a newish attribute to detect this!
https://reviews.llvm.org/rL338464

I can confirm that this case is detected with clang-10 and the following change to span.h:
```diff
-    constexpr Span(V&& v) noexcept : m_data(v.data()), m_size(v.size()) {}
+    constexpr Span(V&& v __attribute__((lifetimebound))) noexcept : m_data(v.data()), m_size(v.size()) {}
```
A new warning is given:
```
spantest.cpp:12:24: warning: temporary whose address is used as value of local variable 'sp' will be destroyed at the end of the full-expression [-Wdangling]
    Span<const int> sp(std::vector<int>{1, 2, 3});
                       ^~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```
It currently warns about several uses in #13062.
I'm going through them now to see if they're false-positives.

Reply:
@theuni Interesting!

I expect that there is one particular false positive: when you're converting a temporary *Span* to a compatible one. We could introduce a separate constructor for that, which then doesn't get the attribute. On the other hand, perhaps that interferes with useful detection too.

Reply:
@sipa Yes, that seems to be the bulk of them.

I can't figure out why this _isn't_ UB (from ParseScript()):
```c++
    if (Func("pkh", expr)) {
        auto pubkey = ParsePubkey(key_exp_index, expr, ctx != ParseScriptContext::P2WSH, out, error);
```
Remember that ```Func()``` takes a ```std::string```, not a c string. So in the example above, presumably ```expr``` ends up pointing to somewhere inside the temporary string's memory after ```Func()``` returns. As you [pointed out](https://github.com/bitcoin/bitcoin/pull/18468#discussion_r443000807), for string literals that's the lifetime of the program. But in this case, I think we're relying on the fact that std::string's .data() returns that exact pointer. Is that guaranteed to be the case?

Reply:
> We could introduce a separate constructor for that, which then doesn't get the attribute. On the other hand, perhaps that interferes with useful detection too.

Indeed giving Span a (default) move ctor causes all current ```lifetimebound``` false-positives to go away, but still warns when intentionally inserting your UB example.

Reply:
Discussion on this moved to #19387.